Adapt InternalTestCluster to auto adjust `minimum_master_nodes`

bleskes opened the pull request:
#20960 removed `LocalDiscovery` and we now use `ZenDiscovery` in all our tests. To keep cluster forming fast, we are using a `MockZenPing` implementation which uses static maps to return instant results making master election fast. Currently, we don't set `minimum_master_nodes` causing the occasional split brain when starting multiple nodes concurrently and their pinging is so fast that it misses the fact that one of the node has elected it self master. To solve this, `InternalTestCluster` is modified to behave like a true cluster and manage and set `minimum_master_nodes` correctly with every change to the number of nodes.

Tests that want to manage the settings themselves can opt out using a new `autoMinMasterNodes` parameter to the `ClusterScope` annotation. 

Having `min_master_nodes` set means the started node may need to wait for other nodes to be started as well. To combat this, we set `discovery.initial_state_timeout` to `0` and wait for the cluster to form once all node have been started.

With this change in place, I think we can/should do a couple of follow ups:
1) Remove async node starts. Now we can just start the nodes one by one (using `initial_state_timeout=0`) and wait for them to join at the end. This will simplify things.
2) ~~`MockZenPing` makes pinging fast. If a node waits on other nodes to be started, it can busy-ping generating lots of logs. I have some ideas of how to tackle that but I hope that removing async  multi node starting (meaning we can reduce the time a node waits by first creating all of them and then starting them)  will be enough to not require this.~~. This is addressed by this PR as well.

nik9000 commented:
> Tests that want to manage the settings themselves can opt out using a new autoMinMasterNodes parameter to the ClusterScope annotation.

Maybe better to look and see if the setting is set and then do something with it? Or if you want to be more explicit then `autoMinMasterNodes` can check for the setting and fail if it is set?

bleskes commented:
> Maybe better to look and see if the setting is set and then do something with it? Or if you want to be more explicit then autoMinMasterNodes can check for the setting and fail if it is set?

I actually starting by looking at settings. It ended up being more complex - there is ambiguity - what happens if the first node doesn't have it and a follow up node does?

bleskes commented:
PS - I'm working on adding internal cluster tests , but I think we can start reviewing the main code

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -844,21 +879,39 @@ void closeNode() throws IOException {
-        void restart(RestartCallback callback, boolean clearDataIfNeeded) throws Exception {
-            assert callback != null;
-            resetClient();
+        /***
```
Extra `*`

bleskes replied:
gone.

nik9000 commented:
> I actually started with looking at settings. It ended up being more complex - there is ambiguity - what happens if the first node doesn't have it and a follow up node does?

Cool. I see you already have the assertion that no node can have it if you are using auto minimum-master-nodes. Which seems right to me.

nik9000 commented:
I think this is big enough that I'd another reviewer. I'm happy to give it the first shot but I don't feel qualified to catch enough of the implications.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1190,7 +1255,9 @@ public synchronized ClusterService clusterService(@Nullable String node) {
         NodeAndClient randomNodeAndClient = getRandomNodeAndClient(predicate);
-        assert randomNodeAndClient != null;
+        if (randomNodeAndClient == null) {
+            assert randomNodeAndClient != null;
```
Leftover from a debugging session?

bleskes replied:
indeed. Removed.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1435,15 +1451,97 @@ public void rollingRestart() throws Exception {
+            validateClusterFormed();
+        }
+        if (updateMinMaster) {
+            updateMinMasterNodes(0);
```
Should this increment?

nik9000 replied:
Oh, no, it doesn't have to because `updateMinMasterNodes` takes into account the running nodes and the node is running again.

nik9000 replied:
I'm not really sure I like passing in a delta to `updateMinMasterNodes`. I understand why you do it but maybe the number of master nodes is a better thing to pass even if it is more verbose.

I think it'd be more clear if this method looked like:
```java
int currentMasterNodes = getMasterNodeCount();
if (updateMinMaster) {
  // The master is going down but we want the cluster to function as though enough master nodes are present
  updateMinMasterNodes(currentMasterNodes - 1);
}
restart it
wait for it to be up if needed
if (updateMinMaster) {
  // The master is back so we can tell nodes about it.
  updateMinMasterNodes(currentMasterNodes);
}
```

I just get confused because when I work in deltas `-1` should balance with `+1` but when I work in absolutes I don't have to balance anything.

bleskes replied:
agreed. I moved it to an absolute number

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -438,15 +438,19 @@ public static int randomInt(int max) {
+    }
+
+    /** Pick a random object from the given collection. */
+    public static <T> T randomFrom(Random random, Collection<T> collection) {
```
I'm not sure this is related.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -527,10 +528,12 @@ protected final void afterInternal(boolean afterClass) throws Exception {
+                        final Map<String, String> transientSettings =  new HashMap<>(metaData.transientSettings().getAsMap());
+                        // this is set by the test infra
+                        transientSettings.remove(ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING.getKey());
+                        assertThat("test leaves transient cluster metadata behind: " + transientSettings, transientSettings.size(), equalTo(0));
```
`hasSize(0)` or `empty()` produce nicer error messages.

bleskes replied:
yeah, I was thinking about this too. I don't think we necessarily need to care if a test leaves settings behind.

bleskes replied:
sure thing. changed.

nik9000 commented on core/src/test/java/org/elasticsearch/recovery/RelocationIT.java:
```diff
@@ -424,22 +424,23 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
-        logger.info("blue nodes: {}", blueFuture.get());
-        logger.info("red nodes: {}", redFuture.get());
+        Settings[] nodeSettings = Stream.concat(
+            Stream.generate(() -> Settings.builder().put("node.attr.color", "blue").build()).limit(halfNodes),
```
I've not seen this one. I keep wanting to reach for `IntStream.range().mapToObject()` for this kind of thing but I like this better.

bleskes replied:
hehe. I just google every time until I see something I can live with :)

bleskes commented on core/src/test/java/org/elasticsearch/cluster/MinimumMasterNodesIT.java:
```diff
@@ -273,12 +273,14 @@ public void testDynamicUpdateMinimumMasterNodes() throws Exception {
-        assertThat(clusterHealthResponse.isTimedOut(), equalTo(false));
+        logger.info("--> start second node and wait for it to join");
+        internalCluster().startNode(settings);
+        ensureClusterSizeConsistency();
```
These changes are needed because the test start two nodes concurrently without setting min master nodes. This can cause split brain - this can happen normally too but switching to MockZenPings (to speed up tests) makes it much more likely. Instead of running the risk I changed the test to start nodes one at a time.

javanna replied:
does it make sense to add a comment on that?

bleskes commented on core/src/test/java/org/elasticsearch/gateway/RecoveryFromGatewayIT.java:
```diff
@@ -332,48 +332,39 @@ public void testLatestVersionLoaded() throws Exception {
-                    for (int i = 0; i < 10; i++) {
-                        assertHitCount(client.prepareSearch().setSize(0).setQuery(matchAllQuery()).execute().actionGet(), 3);
-                    }
+        logger.info("--> one node is closed - start indexing data into the second one");
```
the test starts two nodes, shut down one and while it's down makes changes to the other. The it shuts the other one down and brings both of the up together. I had some trouble with the full restart in the early versions of the change and I decided to simplify the test to make it cleared what it does

bleskes commented on core/src/test/java/org/elasticsearch/recovery/RelocationIT.java:
```diff
@@ -424,22 +424,23 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
+        Settings[] nodeSettings = Stream.concat(
+            Stream.generate(() -> Settings.builder().put("node.attr.color", "blue").build()).limit(halfNodes),
+            Stream.generate(() -> Settings.builder().put("node.attr.color", "red").build()).limit(halfNodes)
+            ).toArray(Settings[]::new);
```
it was clearer but the problem with the old way is that the test cluster had no idea how many nodes are coming.  That was annoying as you can not go directly from a 2 nodes cluster to a 4 nodes cluster (a 2 nodes cluster won't let you update min master nodes to 3) so you'll need to start those nodes one by one. I found the current "start 4 nodes with a bang" simpler compared to that

bleskes commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -1518,6 +1521,8 @@ private void postIndexAsyncActions(String[] indices, List<CountDownLatch> inFlig
          */
         boolean supportsDedicatedMasters() default true;
 
+        boolean autoMinMasterNodes() default true;
```
yes!

bleskes commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -777,6 +808,10 @@ public int nodeAndClientId() {
             return nodeAndClientId;
         }
 
+        public boolean isMasterNode() {
```
yeah - this is always a good one. It will take a much bigger change. For example, the setting is called `NODE_MASTER_SETTING` - i opted to use the short lingo

javanna replied:
I know I know. I just thought that it would add clarity though, cause is MasterNode is really ambiguous. I don't think it breaks any convention if we use the right terminology when adding new methods, especially in test code.

nik9000 replied:
I'd prefer to use "isMasterEligible" here too. I mean, I'd prefer it everywhere but I think it'd be ok to start using it here.

bleskes replied:
OK. I'll change.

bleskes commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -972,13 +1027,17 @@ private synchronized void reset(boolean wipeData) throws IOException {
-            publishNode(nodeAndClient);
+            settings.put(Node.NODE_MASTER_SETTING.getKey(), true);
+            settings.put(Node.NODE_DATA_SETTING.getKey(), false);
+            settings.put(INITIAL_STATE_TIMEOUT_SETTING.getKey(), "0s"); // we wait at the end
```
normally a node will wait for it's initial state to be set (by either joining an existing master or electing itself). However since we now set the min master nodes setting, the need will not being to do so alone and will jest end up waiting 30s. This setting makes the node finish starting immediately and we check for cluster forming at the end.

javanna replied:
do you think it makes sense to add a comment about this?

bleskes commented on test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java:
```diff
@@ -636,7 +636,7 @@ public static void assertVersionSerializable(Version version, Streamable streama
          */
         NamedWriteableRegistry registry;
-        if (ESIntegTestCase.isInternalCluster()) {
+        if (ESIntegTestCase.isInternalCluster() && ESIntegTestCase.internalCluster().size() > 0) {
```
The assertion in `ESIntegTestCase.internalCluster().getInstance(NamedWriteableRegistry.class);` blows up if there are no nodes in the cluster. I stumbled on this while debugging

nik9000 replied:
This is mostly used to make sure that all communications between the embedded nodes serialize at some random version. Since you can only communicate if there is a node to communicate with, I'm somewhat surprised you got here. The condition makes sense and I'm fine with it, but it seems curious that you'd make it here. I wonder if there is some race condition that never comes up in CI but you can trigger it by stepping through certain code.

Either way, I'm ok with the addition here.

bleskes replied:
I don't remember any more to be honest but it was a test managed its own cluster that wasn't available here.

bleskes commented:
@nik9000 @javanna this for the review. I know it's a tough one. I addressed all your comments. I also had enough of the logs that come out of the `MockZenPings` busy spinning, so I addressed this as well. Can you take another look?

bleskes commented:
retest this please

javanna commented on core/src/test/java/org/elasticsearch/recovery/RelocationIT.java:
```diff
@@ -351,7 +351,8 @@ public void testCancellationCleansTempFiles() throws Exception {
         client().admin().indices().prepareCreate(indexName)
                 .setSettings(Settings.builder().put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1, IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0)).get();
 
-        internalCluster().startNodesAsync(2).get();
```
it seems like we removed many usages of `startNodesAsync`. Does it still make sense to have that async method if it causes problems with how the test infra is setup by default?

nik9000 replied:
I believe @bleskes talked about removing it entirely as a followup.

javanna commented on core/src/test/java/org/elasticsearch/recovery/RelocationIT.java:
```diff
@@ -424,22 +425,23 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
+        Settings[] nodeSettings = Stream.concat(
+            Stream.generate(() -> Settings.builder().put("node.attr.color", "blue").build()).limit(halfNodes),
+            Stream.generate(() -> Settings.builder().put("node.attr.color", "red").build()).limit(halfNodes)
+            ).toArray(Settings[]::new);
```
I see, ok. Starting all nodes with one single call is what I had missed

javanna commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -1519,6 +1524,12 @@ private void postIndexAsyncActions(String[] indices, List<CountDownLatch> inFlig
         boolean supportsDedicatedMasters() default true;
 
         /**
+         * Normally the cluster will auto manage the {@link ElectMasterService#DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING} automatically
```
I would slightly rephrase to:
The cluster automatically manages the {@link ElectMasterService#DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING} by default as nodes are started and stopped. Set this to false to manage the setting manually.

bleskes replied:
sure thing

javanna commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -844,21 +874,40 @@ void closeNode() throws IOException {
-            createNewNode(newSettings);
-            startNode();
+            createNewNode(newSettings.build());
+            // make sure cashed client points to new node
```
cached?

nik9000 replied:
++

javanna commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1534,19 +1634,54 @@ public synchronized String startNode(Settings.Builder settings) {
+     *
+     * @param eligableMasterNodeCount the number of master eligable nodes to use as basis for the min master node setting
+     */
+    private int updateMinMasterNodes(int eligableMasterNodeCount) {
```
s/eligable/eligible

bleskes replied:
👍

javanna commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1534,19 +1634,54 @@ public synchronized String startNode(Settings.Builder settings) {
+     */
+    private int updateMinMasterNodes(int eligableMasterNodeCount) {
+        assert autoManageMinMasterNodes;
+        final int currentMasters = getMasterNodesCount();
```
nit: maybe we don't even need this variable?

bleskes replied:
no longer needed indeed. There used to be much more logic around it.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1534,19 +1634,54 @@ public synchronized String startNode(Settings.Builder settings) {
+        assert autoManageMinMasterNodes;
+        final int currentMasters = getMasterNodesCount();
+        final int minMasterNodes = getMinMasterNodes(eligableMasterNodeCount);
+        if (currentMasters == 0) {
```
maybe we could simply do if (getMasterNodesCount>0) rather than having an empty if block?

bleskes replied:
yep. After the latest clean up of other clauses this is silly.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1613,27 +1752,28 @@ public synchronized String startDataOnlyNode(Settings settings) {
+            defaultMinMasterNodes = updateMinMasterNodes(getMasterNodesCount() + mastersDelta);
+        } else {
+            defaultMinMasterNodes = -1;
+        }
```
this code block is very similar to something I already saw above, shall we factor it out?

bleskes replied:
I'm going to remove all of this ... rather not change more :)

javanna commented on test/framework/src/test/java/org/elasticsearch/test/test/InternalTestClusterTests.java:
```diff
@@ -343,6 +403,59 @@ public Settings transportClientSettings() {
         }
+    }
+
+    public void testTwoNodeCluster() throws Exception {
```
++ for this test

nik9000 commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -669,11 +669,11 @@ private Node stop() {
         injector.getInstance(SnapshotShardsService.class).stop();
         // stop any changes happening as a result of cluster state changes
         injector.getInstance(IndicesClusterStateService.class).stop();
+        injector.getInstance(Discovery.class).stop();
```
If the position of this is important I'd leave a comment about why. Even if it is just to stop spammy log messages.

bleskes replied:
added.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -527,10 +529,13 @@ protected final void afterInternal(boolean afterClass) throws Exception {
+                        assertThat("test leaves persistent cluster metadata behind: " + persistent, persistent.size(), equalTo(0));
+                        final Map<String, String> transientSettings =  new HashMap<>(metaData.transientSettings().getAsMap());
+                        // this is set by the test infra
+                        transientSettings.remove(ElectMasterService.DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING.getKey());
```
Should you only remove this if the annotation asks for this to be automatically managed?

bleskes replied:
good idea

bleskes commented:
merged from master again to bring in percolator fixes.

bleskes commented:
Thx @javanna @nik9000 for the lengthy reviews. I will give it some time on master before backporting.
